File: src/basecamp/hooks/file_length.py
"""PostToolUse hook: warn (never block) when a source file grows past the cap.

Fires after a ``Write``/``Edit`` succeeds, so the file is already on disk. This
reads it, counts its lines, and — for a *source* file over the cap — returns a
non-blocking advisory (``hookSpecificOutput.additionalContext``) telling the agent
to review it. It emits no ``decision`` field, so the tool result is untouched and
the write stands; the cap's rationale lives in the engineering doctrine. Anything
unexpected (missing file, non-source suffix, odd payload) returns ``None`` and the
hook stays silent — the primary channel is the doctrine, this is only a nudge.
"""

from __future__ import annotations

import json
from collections.abc import Mapping
from pathlib import Path
from typing import Any

#: Universal line cap for source files. Split along responsibility seams past this.
LINE_CAP = 500

#: Extensions treated as source code. Data, docs, lockfiles, and config are exempt.
_SOURCE_SUFFIXES = frozenset(
    {
        ".py",
        ".pyi",
        ".ts",
        ".tsx",
        ".js",
        ".jsx",
        ".mjs",
        ".cjs",
        ".go",
        ".rs",
        ".java",
        ".kt",
        ".kts",
        ".scala",
        ".swift",
        ".rb",
        ".php",
        ".lua",
        ".jl",
        ".c",
        ".h",
        ".cpp",
        ".cc",
        ".cxx",
        ".hpp",
        ".hh",
        ".cs",
        ".sh",
        ".bash",
        ".zsh",
        ".sql",
    }
)


def _target_path(payload: Mapping[str, Any]) -> Path | None:
    """Resolve the written file's absolute path from the hook payload."""
    tool_input = payload.get("tool_input")
    if not isinstance(tool_input, dict):
        return None
    raw = tool_input.get("file_path")
    if not isinstance(raw, str) or not raw:
        return None
    path = Path(raw)
    if not path.is_absolute():
        cwd = payload.get("cwd")
        if isinstance(cwd, str) and cwd:
            path = Path(cwd) / path
    return path


def handle_file_length(payload: Mapping[str, Any]) -> str | None:
    """Return a non-blocking advisory when a written source file exceeds the cap."""

    if payload.get("tool_name") not in ("Write", "Edit", "MultiEdit"):
        return None
    path = _target_path(payload)
    if path is None or path.suffix.lower() not in _SOURCE_SUFFIXES:
        return None
    try:
        lines = len(path.read_text(encoding="utf-8").splitlines())
    except (OSError, UnicodeDecodeError):
        return None
    if lines <= LINE_CAP:
        return None
    context = (
        f"basecamp file-length check: {path.name} is now {lines} lines, over the "
        f"{LINE_CAP}-line cap for source files. Files past the cap should be split "
        "along responsibility seams into focused modules (see the engineering doctrine)."
    )
    return json.dumps(
        {
            "hookSpecificOutput": {
                "hookEventName": "PostToolUse",
                "additionalContext": context,
            }
        }
    )

File: omp/engineering/file-length.ts
import { readFileSync } from "node:fs";
import * as path from "node:path";
import type { ExtensionAPI } from "@oh-my-pi/pi-coding-agent";

const DEFAULT_SOURCE_LINE_LIMIT = 500;
const FILE_LENGTH_REMINDER_TYPE = "basecamp-file-length-reminder";

const SOURCE_SUFFIXES = new Set([
	".bash",
	".c",
	".cc",
	".cjs",
	".cpp",
	".cs",
	".css",
	".cxx",
	".go",
	".h",
	".hh",
	".hpp",
	".htm",
	".html",
	".java",
	".jl",
	".js",
	".jsx",
	".kt",
	".kts",
	".lua",
	".mjs",
	".php",
	".py",
	".pyi",
	".rb",
	".rs",
	".scala",
	".sh",
	".sql",
	".swift",
	".ts",
	".tsx",
	".zsh",
]);

const SOURCE_LINE_LIMITS = new Map<string, number>([
	[".htm", 350],
	[".html", 350],
	[".ts", 350],
	[".tsx", 350],
	[".bash", 400],
	[".sh", 400],
	[".zsh", 400],
	[".sql", 800],
]);

function lineLimit(filePath: string): { limit: number; suffix: string } | null {
	const suffix = path.extname(filePath).toLowerCase();
	if (!SOURCE_SUFFIXES.has(suffix)) return null;
	return { limit: SOURCE_LINE_LIMITS.get(suffix) ?? DEFAULT_SOURCE_LINE_LIMIT, suffix };
}

function lineCount(content: string): number {
	if (content === "") return 0;
	// Counting only newlines matches editor gutters and the repository's hard checker.
	return content.split("\n").length - (content.endsWith("\n") ? 1 : 0);
}

function reminderContent(filePath: string, lines: number, limit: number, suffix: string): string {
	return (
		"<system-reminder>\n" +
		`File-length reminder: ${JSON.stringify(filePath)} is now ${lines} lines, over the ${limit}-line cap for ${suffix} source files. ` +
		"The edit succeeded; this is advisory. Split the file along genuine responsibility seams into focused modules. " +
		"Do not compress formatting or create continuation files merely to satisfy the cap, and follow any tighter project-specific limit.\n" +
		"</system-reminder>"
	);
}

export default function registerFileLengthReminder(omp: ExtensionAPI): void {
	const remindedPaths = new Set<string>();
	const reset = (): void => remindedPaths.clear();

	omp.on("session_start", reset);
	omp.on("agent_end", (event) => {
		if (event.willContinue !== true) reset();
	});

	omp.on("tool_result", (event, ctx) => {
		if (event.isError) return;
		if (event.toolName !== "edit" && event.toolName !== "write") return;

		const rawPath = event.input.path;
		if (typeof rawPath !== "string" || rawPath === "") return;

		try {
			const resolvedPath = path.isAbsolute(rawPath) ? path.resolve(rawPath) : path.resolve(ctx.cwd, rawPath);
			const policy = lineLimit(resolvedPath);
			if (!policy) return;

			const lines = lineCount(readFileSync(resolvedPath, "utf8"));
			if (lines <= policy.limit) {
				remindedPaths.delete(resolvedPath);
				return;
			}
			if (remindedPaths.has(resolvedPath)) return;

			omp.sendMessage(
				{
					customType: FILE_LENGTH_REMINDER_TYPE,
					content: reminderContent(resolvedPath, lines, policy.limit, policy.suffix),
					display: false,
				},
				{ deliverAs: "steer" },
			);
			remindedPaths.add(resolvedPath);
		} catch {
			// Advisory failures must never affect a successful edit.
		}
	});
}
